boards/arm/pico_pi_m4/doc/pico_pi_p4.rst

```diff
@@ -0,0 +1,300 @@
+
+
+References
+==========
+
+- `i.MX 7 Dual Reference Manual`_ from page 182 (section 2.1.2 and 2.1.3)
+- `Toradex Wiki`_
+
+
+At compilation time you have to choose which RAM will be used. This
+configuration is done in the file ``boards/arm/pico_pi_m4/pico_pi_m4.dts`` with
+"zephyr,flash" (when CONFIG_XIP=y) and "zephyr,sram" properties. The available
+configurations are:
+
+.. code-block:: none
+
+   "zephyr,flash"
+   - &ddr_code
+   - &tcml_code
+   - &ocram_code
+   - &ocram_s_code
+   - &ocram_pxp_code
+   - &ocram_epdc_code
+
+   "zephyr,sram"
+   - &ddr_sys
+   - &tcmu_sys
+   - &ocram_sys
+   - &ocram_s_sys
+   - &ocram_pxp_sys
+   - &ocram_epdc_sys
+
+
+Below you will find the instructions to load and run Zephyr on M4 from A7 using
+u-boot.
+
+Connect both micro USB interfaces into the PC. This is the A7 console and the
+UART6 in the Edison connector is M4 console for Zephyr with both configured to work at 115200 8N1.
+The USB interface is used to power the CPU and IO boards and is connected
+to the USB OTG interface of the i.MX7D.
+
+After powering up the platform stop the u-boot execution on the A7 core and
+expose the eMMC as mass storage with the following command in the u-boot
+prompt: ``ums 0 mmc 0``. Copy the compiled zephyr.bin to the first FAT
+partition and remove the mounted device on the PC by issuing a "Ctrl+C" in the
+u-boot prompt.
+Set the u-boot environment variables and run the zephyr.bin from the
+appropriated memory configured in the Zephyr compilation:
+
+.. code-block:: console
+
+   setenv bootm4 'fatload mmc 0:1 $m4addr $m4fw && dcache flush && bootaux $m4addr'
+   # TCML
+   setenv m4tcml 'setenv m4fw zephyr.bin; setenv m4addr 0x007F8000'
+   setenv bootm4tcml 'run m4tcml && run bootm4'
+   run bootm4tcml
+   # TCMU
+   setenv m4tcmu 'setenv m4fw zephyr.bin; setenv m4addr 0x00800000'
+   setenv bootm4tcmu 'run m4tcmu && run bootm4'
+   run bootm4tcmu
+   # OCRAM
+   setenv m4ocram 'setenv m4fw zephyr.bin; setenv m4addr 0x00900000'
+   setenv bootm4ocram 'run m4ocram && run bootm4'
+   run bootm4ocram
+   # OCRAM_S
+   setenv m4ocrams 'setenv m4fw zephyr.bin; setenv m4addr 0x00180000'
+   setenv bootm4ocrams 'run m4ocrams && run bootm4'
+   run bootm4ocrams
+   # DDR
+   setenv m4ddr 'setenv m4fw zephyr.bin; setenv m4addr 0x80000000'
+   setenv bootm4ddr 'run m4ddr && run bootm4'
+   run bootm4ddr
+
+
+Debugging
+=========
+
+Download and install `J-Link Tools`_ and `NXP iMX7D Connect CortexM4.JLinkScript`_.
+
+To run Zephyr Binary using J-Link, create the following script to get the
+Program Counter and Stack Pointer from ``zephyr.bin``.
+
+get-pc-sp.sh:
+.. code-block:: console
+
+   #!/bin/sh
+
+   firmware=$1
+
+   pc=$(od -An -N 8 -t x4 $firmware | awk '{print $2;}')
+   sp=$(od -An -N 8 -t x4 $firmware | awk '{print $1;}')
+
+   echo pc=$pc
+   echo sp=$sp
+
+
+Get the SP and PC from firmware binary: ``./get-pc-sp.sh zephyr.bin``
+.. code-block:: console
```

Comment:
add a blank line before this code-block directive

Reply:
It's done